pkg/provider/bitwarden/client.go

```diff
@@ -0,0 +1,311 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package bitwarden
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/kube-openapi/pkg/validation/strfmt"
+
+	esv1beta1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1beta1"
+	"github.com/external-secrets/external-secrets/pkg/utils"
+)
+
+var (
+	errBadCertBundle = "caBundle failed to base64 decode: %w"
+)
+
+const (
+	// NoteMetadataKey defines the note for the pushed secret.
+	NoteMetadataKey = "note"
+)
+
+// PushSecret will write a single secret into the provider.
+// Note: We will refuse to overwrite ANY secrets, because we can never be completely
+// sure if it's the same secret we are trying to push. We only have the Name and the value
+// could be different. Therefore, we will always create a new secret. Except if, the value
+// the key, the note, and organization ID all match.
```

Comment:
I think this is really hard to use if secrets are being rotated in the cluster and eventually pushed to bitwarden. We end up with lots of secrets, and no one knows which one is the right one.

Is there a possibility to identify that a secret is "ours" ?
With other providers we attempt to identify secrets based on labels/annotations where available and ensure that we only modify secrets that we own. Maybe we can build something similar by (mis-?)using the `note` to store data that indicates ownership by ESO and other relevant metadata?

Reply:
That's a good idea to misuse notes. There isn't anything else sadly. There is no metadata section for a secret.

The only concern would be if someone would want to add more notes and we would have to somehow parse that and figure out if our thing is inside that note too. 

But maybe it's fine. I'll play around with that.

Reply:
The reason I was reluctant here and so cautious is that names are not unique not even in the same project. So unless someone defines a UUID ( which is not great UX but that's how the Bitwarden community wants us to use Bitwarden... ) there is a very real risk of overwriting something that you aren't supposed to.

I could restrict the usage to UUID... But people already said that it's not nice UX. But maybe being precise trumps UX. WDYT?

Reply:
Actually:
>I think this is really hard to use if secrets are being rotated in the cluster and eventually pushed to bitwarden. We end up with lots of secrets, and no one knows which one is the right one.

So, just to be clear. You mean, if people change the complete KEY + VALUE of a secret? Because if only the value is change we'll just overwrite it. But if the key changes, the previous one will get left behind, right? That's what you are pointing out here?

Reply:
Maybe it's enough to just put something into the notes? Something like:

```
=== ESO ===
Secret managed by external-secrets operator.
=== ESO ===
```

?

Reply:
I haven't tried it out yet, from the comment i read:
If a secret value changes in the cluster (which is referenced from a PushSecret), then ESO will create a new secret with the same name but new uuud in Bitwarden. 

So, new value => new secret in bitwarden. 

If that is not the case ignore the comment :)

Reply:
ah, yes. That's not the case. Perhaps, I need to rephrase that sentence. If orgID + projectID + Note + Key all equal but the VALUE is different, we will update the existing secret. The Note is optional, it can be empty. But then it must be empty when checking again. ( If someone happens to update the secret by hand, we will basically create a new one with the note set to the specific note defined by the PushSecret ).

Reply:
@moolen Can I resolve this conversation? :)

Reply:
Yeah, sure. I'm struggling to find time for another review round   ;)

Reply:
No worries! I know you have better priorities right now. 🤣😍